# LLM Fundamentals

---

## 1. LLM Mental Model

Before writing any code, we need to understand the high-level path that text follows inside a Large Language Model (LLM).

A user sees:

```text
"I love machine learning."
```

But the model does not process this sentence as human-readable text.

Instead, the text is transformed into numerical representations before the Transformer can process it.

The overall pipeline is:

```text
User Text
   ↓
Tokenizer
   ↓
Tokens
   ↓
Token IDs
   ↓
Embeddings
   ↓
Position Information
   ↓
Decoder-Only Transformer
   ↓
Hidden States
   ↓
Language Modeling Head
   ↓
Logits
   ↓
Next-Token Prediction
```

For the Local AI Assistant, this pipeline is the foundation of the entire inference process.

The first part of this pipeline is:

```text
Text
 ↓
Tokenization
 ↓
Token IDs
 ↓
Embeddings
 ↓
Transformer Input
```

The important mental model is:

> **Language enters the model as numbers, not as raw human-readable text.**

This means that an LLM must first transform language into representations that neural network operations can process.

---

## 2. What Exactly Is a Token?

A **token** is a unit produced by the tokenizer.

A token is not necessarily:

```text
1 token = 1 word
```

and it is not necessarily:

```text
1 token = 1 character
```

A token can represent:

- A whole word
- A part of a word
- Punctuation
- Whitespace-related pieces
- A number
- A symbol
- A special token

Therefore:

```text
1 token ≠ 1 word
1 token ≠ 1 character
```

The exact tokenization behavior depends on the tokenizer associated with the model.

For example, a word such as:

```text
unbelievable
```

may be represented using multiple subword pieces:

```text
un | believ | able
```

This is only an illustrative example.

The exact token boundaries depend on the tokenizer being used.

The important concept is:

```text
Text
 ↓
Tokenizer
 ↓
Token sequence
```

A token can therefore represent a reusable piece of language rather than necessarily representing an entire word.

---

## 3. Why Tokenization Exists

Neural networks operate on numerical tensors.

Raw text such as:

```text
"Machine Learning"
```

cannot simply be passed directly into the model as a string.

The text must first be converted into a sequence of tokens:

```text
Raw Text
   ↓
Tokenization
   ↓
Token Sequence
```

Then the token sequence is converted into integer IDs:

```text
Token Sequence
   ↓
Token IDs
```

The model can then operate on these numerical representations.

The complete simplified flow is:

```text
Text
 ↓
Tokens
 ↓
Token IDs
 ↓
Embeddings
 ↓
Transformer
```

Tokenization therefore acts as the bridge between human language and the numerical input representation used by the model.

---

## 4. Example: Tokenization

Consider:

```text
"Machine learning is useful."
```

A tokenizer could conceptually produce something like:

```text
["Machine", " learning", " is", " useful", "."]
```

The exact result is model-dependent.

One important observation is that spaces may be represented as part of token boundaries.

For example:

```text
"Machine"
" learning"
```

may be treated as different token pieces.

Do not memorize a specific tokenization output as universal.

The correct mental model is:

```text
Same Text
   ↓
Different Tokenizer
   ↓
Possibly Different Tokens
```

This means that tokenization is a property of the tokenizer/model pair, not a universal property of language.

---

## 5. Why Tokens Are Useful

Subword tokenization allows the vocabulary to reuse pieces instead of storing every possible word independently.

For example, a tokenizer does not need a separate vocabulary entry for every imaginable variation of a word.

Instead, it can reuse smaller pieces.

This helps the model handle:

- Rare words
- New words
- Morphological variations
- Technical terminology
- Source code
- Mixed-language text

Conceptually:

```text
Many Possible Words
        ↓
Reusable Token Pieces
        ↓
Efficient Vocabulary
```

This is one reason tokenization is such an important part of modern LLM systems.

---

## 6. Tokenization and Vocabulary

The tokenizer works together with a **vocabulary**.

Conceptually:

```text
Vocabulary
────────────────
token A
token B
token C
token D
...
token N
```

Each vocabulary item is associated with an integer identifier.

For example, conceptually:

```text
"machine"   → 1234
"learning"  → 5678
"."         → 42
```

These numbers are only illustrative.

The actual IDs depend entirely on the tokenizer and model.

The relationship is:

```text
Token
   ↓
Vocabulary Lookup
   ↓
Integer ID
```

The vocabulary therefore defines the set of token units that the tokenizer can map to IDs.

---

## 7. Token IDs

After tokenization, the token sequence is mapped to integer IDs.

Conceptually:

```text
Tokens
   ↓
Integer IDs
```

For example:

```text
["I", "love", "AI"]
```

could conceptually become:

```text
[101, 532, 7281]
```

Again, these values are illustrative.

The important relationship is:

```text
Token
   ↓
Token ID
```

The token ID is an integer identifier assigned to a token in the tokenizer vocabulary.

These IDs are what the model uses as indices into its embedding representation.

---

## 8. Why Integer IDs?

The model needs a compact numerical representation for each token.

Token IDs provide an efficient way to identify vocabulary entries.

Conceptually:

```text
Token ID
   ↓
Embedding Matrix Lookup
   ↓
Embedding Vector
```

For example:

```text
"machine"
    ↓
Token ID
    ↓
Embedding Lookup
    ↓
Vector
```

An important distinction is:

> **A token ID is not the semantic representation of a token.**

The ID is simply an index.

The learned semantic representation comes from the embedding vectors and later contextual representations produced by the Transformer.

---

## 9. Special Tokens

Modern tokenizers may include special tokens that are used to control or represent structural information.

Examples can include:

```text
BOS
EOS
PAD
EOT
Other model-specific control tokens
```

The exact set and meaning depend on the tokenizer/model family.

Important:

```text
Not every model uses the same special tokens.
```

Special tokens may be used to indicate things such as:

- Start of a sequence
- End of a sequence
- Padding
- End of a turn
- Conversation structure

We should therefore inspect the actual tokenizer rather than assuming a universal set of special tokens.

---

## 10. Embeddings

An **embedding** is a numerical vector representing a token.

Conceptually:

```text
"machine"
    ↓
[0.12, -0.44, 0.91, ..., 0.07]
```

The real vectors contain many dimensions.

We do not manually choose these numbers.

The pretrained model contains learned parameters that map token IDs to embedding vectors.

The simplified flow is:

```text
Token ID
   ↓
Embedding Lookup
   ↓
Embedding Vector
```

This is the stage where token identifiers are transformed into continuous numerical representations that can participate in neural network computation.

---

## 11. Embedding Matrix

The model contains an embedding table, which can be viewed conceptually as a matrix.

Suppose:

```text
V = Vocabulary Size
D = Embedding Dimension
```

Then the embedding matrix has a conceptual shape:

```text
[V, D]
```

A simplified representation is:

```text
                 Embedding Dimension
                         ↓

Token 1      [ ... ... ... ... ]
Token 2      [ ... ... ... ... ]
Token 3      [ ... ... ... ... ]
Token 4      [ ... ... ... ... ]
   ...
Token V      [ ... ... ... ... ]
```

Each row corresponds to a vocabulary item.

A token ID selects one row from this matrix.

Therefore:

```text
Token ID
   ↓
Row Lookup
   ↓
Embedding Vector
```

This is an important bridge between tokenization and neural network computation.

---

## 12. Token vs Token ID vs Embedding vs Hidden State

These four concepts must not be confused.

### Token

A token is the tokenizer-level representation of a piece of text.

Example:

```text
"learning"
```

---

### Token ID

A token ID is the integer identifier assigned to that token.

Example:

```text
5678
```

---

### Embedding

An embedding is the initial learned numerical vector associated with the token ID.

Example:

```text
[0.1, -0.3, 0.8, ...]
```

---

### Hidden State

A hidden state is a context-dependent representation produced after the token representation has passed through Transformer computation.

Conceptually:

```text
Embedding
   ↓
Transformer
   ↓
Hidden State
```

This means:

```text
Token
   ↓
Token ID
   ↓
Embedding
   ↓
Transformer Processing
   ↓
Contextual Hidden State
```

The embedding and hidden state are therefore different objects.

The embedding is an initial learned representation, while the hidden state contains contextual information produced by the Transformer.

---

## 13. Position Matters

Knowing which tokens exist is not enough.

The model must also understand where tokens occur in the sequence.

Consider:

```text
"dog bites man"
```

and:

```text
"man bites dog"
```

They contain the same basic words, but their meanings are different.

Therefore, the model needs a way to represent sequence position.

The general principle is:

```text
Token Identity + Position
           ↓
Contextual Processing
```

The exact method used to represent positional information depends on the model architecture.

Different model families may use different positional mechanisms.


> **The model needs information about both token identity and token position.**

We do not need to memorize a specific positional-encoding implementation yet.

---

## 14. Decoder-Only Transformer

The Local AI Assistant will use a pretrained language model built around a decoder-style Transformer architecture.

At a high level:

```text
Token IDs
   ↓
Token Embeddings
   +
Position Information
   ↓
Transformer Blocks
   ↓
Final Hidden States
   ↓
Language Modeling Head
   ↓
Logits
```

A Transformer block contains several important mechanisms.

At a high level, we should recognize:

```text
Self-Attention
Feed-Forward Network
Normalization
Residual Connections
```

These components work together to transform the input token representations into contextual hidden states.

The exact implementation differs between model families.

The objective is to understand the architecture at a conceptual level rather than study every internal implementation detail.

---

### 14.1 Self-Attention

The core intuition behind self-attention is that a token's representation can incorporate information from other relevant tokens in its context.

Consider:

```text
"The animal didn't cross the road because it was tired."
```

To interpret:

```text
"it"
```

the model needs information from the surrounding context.

Self-attention is one of the main mechanisms that allows contextual relationships between tokens to be modeled.

Conceptually:

```text
Token
  ↓
Looks at relevant context
  ↓
Updated contextual representation
```

For decoder-only autoregressive models, this attention is causal, meaning the model cannot use future tokens that have not yet been generated.

---

### 14.2 Q / K / V Preview

We will study the mathematical and implementation details later.

For now, the mental model is:

```text
Hidden States
      ↓
   Q   K   V
      ↓
 Attention
      ↓
Contextual Representation
```

The general attention equation is:
$$
Attention(Q,K,V)
=
softmax\left(\frac{QK^T}{\sqrt{d_k}}\right)V
$$

Understanding what Q, K, and V represent at a high level is enough.

The deeper self-attention mechanics will be studied.

---

### 14.3 Causal Attention

During autoregressive language modeling, the model must predict the next token using only information that is available so far.

Conceptually:

```text
Token 1
  ↓
can see Token 1
```

```text
Token 2
  ↓
can see Token 1 + Token 2
```

```text
Token 3
  ↓
can see Token 1 + Token 2 + Token 3
```

But:

```text
Token 2
   ✗
cannot see Token 3
```

This causal constraint prevents the model from looking ahead during next-token prediction.

Therefore:

```text
Current Context
      ↓
Predict Next Token
```

rather than:

```text
Current Context + Future Tokens
      ↓
Predict Next Token
```

Causal attention is one of the foundations of decoder-only autoregressive generation.

---

### 14.4 Next-Token Prediction

Suppose the input is:

```text
"I am learning"
```

The model does not conceptually generate an entire paragraph in one single prediction step.

Instead, it produces scores for possible next tokens.

Conceptually:

```text
"AI"          → 0.42
"machine"     → 0.21
"Python"      → 0.08
"about"       → 0.05
...
```

These values are illustrative.

The exact values depend on the model and context.

The important idea is:

```text
Current Tokens
      ↓
Predict Next Token
      ↓
Append Next Token
      ↓
Predict Again
```

For example:

```text
"I am learning"
        ↓
     predict
        ↓
       "AI"
```

Then:

```text
"I am learning AI"
        ↓
     predict
        ↓
   "engineering"
```

And again:

```text
"I am learning AI engineering"
        ↓
     predict
        ↓
       "..."
```

This repeated process is called **autoregressive generation**.

---

### 14.5 Logits

Before converting the model's output scores into probabilities, the model produces **logits**.

The simplified pipeline is:

```text
Transformer
    ↓
Hidden States
    ↓
Language Modeling Head
    ↓
Logits
```

If the vocabulary contains many thousands of possible tokens, the model produces a corresponding score for each possible next token.

Conceptually:

```text
Vocabulary
────────────────
token 1   → score
token 2   → score
token 3   → score
...
token N   → score
```

These raw scores are the logits.

A probability distribution can then be derived from the logits.

Sampling strategies such as:

```text
Temperature
Top-k
Top-p
```

## 15 Context Window

An LLM operates within a limited **context window**.

The context window represents the amount of tokenized context that the model can process as part of a single sequence.

Conceptually:

```text
┌─────────────────────────────┐
│        Context Window       │
│                             │
│ System Instruction          │
│ User Message                │
│ Conversation History        │
│ Current Prompt              │
│ Generated Tokens            │
│                             │
└─────────────────────────────┘
```

The context window is measured in **tokens**, not characters.

This matters because token count affects how much information can fit into the model's available context.

Token count becomes important for:

- Chat history
- RAG
- Long documents
- Memory
- Latency
- Memory usage

The relationship can be simplified as:

```text
Text Length
    ↓
Token Count
    ↓
Available Context
    ↓
Computation / Memory Implications
```

Two prompts may look similar as strings while having very different token counts.

For example:

```text
Prompt A → 100 tokens
Prompt B → 900 tokens
```

From an inference-engineering perspective, these are not equivalent workloads.

The longer sequence generally requires more input processing.

This is why token counting is a critical part of LLM engineering.

The concept of the context window will become especially important later when we work with:

```text
W16 → RAG
W17 → Advanced RAG
W18 → Agents
W19 → Inference Engineering
```

The key mental model is:

$$
Context\ Window = The\ token\ budget\ available\ to\ the\ model\ for\ processing\ the\ current\ sequence.
$$


### Mental Model So Far

At this point, the complete conceptual pipeline is:

```text
Human Text
   ↓
Tokenizer
   ↓
Tokens
   ↓
Token IDs
   ↓
Embedding Lookup
   ↓
Embedding Vectors
   +
Position Information
   ↓
Decoder-Only Transformer
   ↓
Contextual Hidden States
   ↓
Language Modeling Head
   ↓
Logits
```

And later, the generation process becomes:

```text
Logits
   ↓
Probability Distribution
   ↓
Decoding / Sampling
   ↓
Next Token
   ↓
Append Token
   ↓
Generate Again
```

The central idea is:

> **An LLM does not process raw human language directly. Text is first tokenized, mapped to token IDs, transformed into numerical representations, processed through a Transformer, and converted into scores for possible next tokens.**

# 1. Tokenization, Token IDs, and Embeddings

This notebook explores how raw text is transformed into tokens,
token IDs, and model-ready numerical representations.

`Experiments`

- Tokenization
- Token IDs
- Special Tokens
- Token Counting
- English Text
- Arabic Text
- Mixed Arabic + English
- Code
- Numbers and Punctuation
- Padding and Truncation
- Attention Masks
- Chat Templates
- Embedding Inspection
- Tensor Shape Tracking
- Mini Forward Pass

## 1.1 Python Environment Check

In [1]:
# Validate the local Python environment

import sys

print("Python:", sys.version)

Python: 3.13.0 (tags/v3.13.0:60403a5, Oct  7 2024, 09:38:07) [MSC v.1941 64 bit (AMD64)]


### Result Interpretation

The notebook is running on Python 3.13.0.

This confirms that the selected Python kernel is active and available for the Week 15 Day 1 experiments.

### Finding

The project is running on the intended Python environment.

## 1.2 Transformers Installation Check

In [2]:
# Install Hugging Face Transformers

%pip install -U transformers

Note: you may need to restart the kernel to use updated packages.


In [3]:
# Verify the Transformers installation

import transformers

print("Transformers:", transformers.__version__)

d:\Projects\Local AI Assistan\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Transformers: 5.18.0


### Result Interpretation

Transformers 5.18.0 is installed and can be imported successfully.

This confirms that the Hugging Face Transformers library is available in the project environment.

### Finding

The Transformers dependency is correctly installed and ready for LLM tokenizer experiments.

## 1.3 Qwen3.8-27B Tokenizer Loading

In [4]:
# Load the Qwen3.8-27B tokenizer

from transformers import AutoTokenizer, AutoConfig

MODEL_NAME = "Qwen/Qwen3.8-27B"

tokenizer = AutoTokenizer.from_pretrained(
    MODEL_NAME,
    use_fast=True
)

config = AutoConfig.from_pretrained(MODEL_NAME)

print("Tokenizer:", type(tokenizer).__name__)
print("Tokenizer vocab size:", tokenizer.vocab_size)
print("Model vocab size:", config.text_config.vocab_size)
print("Special tokens:", tokenizer.special_tokens_map)

Tokenizer: Qwen3_5Tokenizer
Tokenizer vocab size: 248044
Model vocab size: 248320
Special tokens: {'eos_token': '<|im_end|>', 'pad_token': '<|endoftext|>', 'audio_bos_token': '<|audio_start|>', 'audio_eos_token': '<|audio_end|>', 'audio_token': '<|audio_pad|>', 'image_token': '<|image_pad|>', 'video_token': '<|video_pad|>', 'vision_bos_token': '<|vision_start|>', 'vision_eos_token': '<|vision_end|>'}


### Result Interpretation

The tokenizer for the Qwen3.8-27B model was loaded successfully.

The tokenizer reports a vocabulary size of 248,044 tokens.

### Finding

The project is using the tokenizer associated with the selected Qwen3.8-27B model setup.

## 1.4 Basic English Tokenization

In [5]:
# Tokenize a simple English sentence

text = "I love machine learning."

tokens = tokenizer.tokenize(text)

print("Text:", text)
print("Tokens:", tokens)
print("Token count:", len(tokens))

Text: I love machine learning.
Tokens: ['I', 'Ġlove', 'Ġmachine', 'Ġlearning', '.']
Token count: 5


### Result Interpretation

The sentence:

```text
I love machine learning.
```
was tokenized into five token pieces:
```
I
Ġlove
Ġmachine
Ġlearning
.
```
The Ġ marker represents a tokenizer-specific representation associated with a preceding whitespace boundary.

### Finding

A token is not necessarily equivalent to a word or a character. Token boundaries depend on the tokenizer.

## 1.5 Token IDs 


In [6]:
# Convert tokens into token IDs

encoded = tokenizer(text)

input_ids = encoded["input_ids"]

print("Tokens:", tokens)
print("Token IDs:", input_ids)
print("Token count:", len(input_ids))

Tokens: ['I', 'Ġlove', 'Ġmachine', 'Ġlearning', '.']
Token IDs: [40, 2854, 5484, 6618, 13]
Token count: 5


### Result Interpretation

The five tokens were converted into integer token IDs:

```text
I          → 40
Ġlove      → 2854
Ġmachine   → 5484
Ġlearning  → 6618
.          → 13
```
These integers are vocabulary identifiers used by the model input pipeline.

The numerical value of a token ID does not represent semantic meaning. It functions as an index for retrieving the corresponding token representation.

### Finding

The tokenizer converts discrete token pieces into integer IDs that can be processed by the model.

## 1.6 Token-to-ID Mapping


In [7]:
# Inspect the token-to-ID mapping

for token, token_id in zip(tokens, input_ids):
    print(f"{token!r:15} → {token_id}")

'I'             → 40
'Ġlove'         → 2854
'Ġmachine'      → 5484
'Ġlearning'     → 6618
'.'             → 13


### Result Interpretation

The explicit token-to-ID mapping confirms the relationship between each token and its vocabulary identifier.

The mapping is deterministic for the selected tokenizer and input text.

### Finding

Each token in the encoded sequence corresponds to one integer identifier in the tokenizer vocabulary.

## 1.7 Decode Token IDs Back to Text

In [8]:
# Decode token IDs back into text

decoded_text = tokenizer.decode(
    input_ids,
    skip_special_tokens=False
)

print("Original:", text)
print("Decoded :", decoded_text)

Original: I love machine learning.
Decoded : I love machine learning.


### Result Interpretation

The encoded token IDs were decoded back into text successfully.

```text
Original: I love machine learning.
Decoded : I love machine learning.
```
The decoded text matches the original input for this example.

### Finding

The tokenizer successfully performs the round-trip transformation:
```
Text
 ↓
Token IDs
 ↓
Decoded Text

## 1.8 Character, Word, and Token Counts


In [9]:
# Compare character, word, and token counts

character_count = len(text)
word_count = len(text.split())
token_count = len(input_ids)

print("Text:", text)
print("Character count:", character_count)
print("Word count:", word_count)
print("Token count:", token_count)

Text: I love machine learning.
Character count: 24
Word count: 4
Token count: 5


### Result Interpretation

The input text contains:

```text
Characters: 24
Words: 4
Tokens: 5
```
These values are different because tokenization does not operate directly on natural language words or individual characters.

### Finding

Character count, word count, and token count are distinct measurements. LLM context is measured in tokens rather than words or characters.

## 1.9 Arabic Tokenization


In [10]:
# Tokenize an Arabic sentence

arabic_text = "أنا أحب تعلم الذكاء الاصطناعي."

arabic_tokens = tokenizer.tokenize(arabic_text)
arabic_ids = tokenizer(arabic_text)["input_ids"]

print("Text:", arabic_text)
print("Tokens:", arabic_tokens)
print("Token IDs:", arabic_ids)
print("Character count:", len(arabic_text))
print("Word count:", len(arabic_text.split()))
print("Token count:", len(arabic_ids))

Text: أنا أحب تعلم الذكاء الاصطناعي.
Tokens: ['Ø£', 'ÙĨا', 'ĠØ£ØŃب', 'ĠتعÙĦÙħ', 'ĠاÙĦØ°', 'Ùĥاء', 'ĠاÙĦاص', 'Ø·ÙĨ', 'اعÙĬ', '.']
Token IDs: [67321, 148415, 226577, 198868, 149634, 218326, 196720, 149158, 198240, 13]
Character count: 30
Word count: 5
Token count: 10


### Result Interpretation

The Arabic sentence contains:

```text
Characters: 30
Words: 5
Tokens: 10
```
Several words are represented by multiple token pieces.

This demonstrates that tokenization behavior depends on the language and the tokenizer vocabulary.

### Finding

A small number of words can correspond to a larger number of tokens. Word count alone is not sufficient for estimating an LLM input size.


## 1.10 Arabic Token Inspection


In [11]:
# Inspect Arabic tokens using repr()

for i, (token, token_id) in enumerate(zip(arabic_tokens, arabic_ids)):
    print(f"{i:02d} | {token!r} → {token_id}")

00 | 'Ø£' → 67321
01 | 'ÙĨا' → 148415
02 | 'ĠØ£ØŃب' → 226577
03 | 'ĠتعÙĦÙħ' → 198868
04 | 'ĠاÙĦØ°' → 149634
05 | 'Ùĥاء' → 218326
06 | 'ĠاÙĦاص' → 196720
07 | 'Ø·ÙĨ' → 149158
08 | 'اعÙĬ' → 198240
09 | '.' → 13


### Result Interpretation

Inspecting each token together with its token ID shows how the tokenizer decomposes the Arabic input into smaller pieces.

Some words are split across multiple tokens rather than represented by a single token.

The displayed token strings may use tokenizer-specific byte-level representations, while the token IDs remain the numerical identifiers used by the model pipeline.

### Finding

Token boundaries can occur inside words, and token IDs provide the stable numerical representation used by the tokenizer and model.

## 1.11 Arabic Token ID Decoding

In [12]:
# Decode each Arabic token ID individually

for i, token_id in enumerate(arabic_ids):
    decoded_token = tokenizer.decode(
        [token_id],
        skip_special_tokens=False
    )
    
    print(f"{i:02d} | ID: {token_id:<7} | Decoded: {decoded_token!r}")

00 | ID: 67321   | Decoded: 'أ'
01 | ID: 148415  | Decoded: 'نا'
02 | ID: 226577  | Decoded: ' أحب'
03 | ID: 198868  | Decoded: ' تعلم'
04 | ID: 149634  | Decoded: ' الذ'
05 | ID: 218326  | Decoded: 'كاء'
06 | ID: 196720  | Decoded: ' الاص'
07 | ID: 149158  | Decoded: 'طن'
08 | ID: 198240  | Decoded: 'اعي'
09 | ID: 13      | Decoded: '.'


### Result Interpretation

Each Arabic token ID was decoded individually to reveal the text fragment represented by that ID.

The decoded fragments reconstruct the original sentence when combined in sequence.

### Finding

A complete word does not necessarily correspond to a single token. Multiple token pieces can be combined to represent one word.

## 1.12 Mixed Langue

In [13]:
# Tokenize mixed Arabic and English text

mixed_text = "أتعلم Machine Learning باستخدام Python."

mixed_tokens = tokenizer.tokenize(mixed_text)
mixed_ids = tokenizer(mixed_text)["input_ids"]

print("Text:", mixed_text)
print("Tokens:", mixed_tokens)
print("Token IDs:", mixed_ids)
print("Character count:", len(mixed_text))
print("Word count:", len(mixed_text.split()))
print("Token count:", len(mixed_ids))

Text: أتعلم Machine Learning باستخدام Python.
Tokens: ['أت', 'عÙĦÙħ', 'ĠMachine', 'ĠLearning', 'ĠباستخداÙħ', 'ĠPython', '.']
Token IDs: [160409, 171519, 12588, 20271, 179677, 12654, 13]
Character count: 39
Word count: 5
Token count: 7


### Result Interpretation

The mixed-language input:

```text
أتعلم Machine Learning باستخدام Python.
```

contains:

```text
Character count: 39
Word count: 5
Token count: 7
```

The tokenizer produced a single sequence containing Arabic text, English technical terms, and punctuation.

The observed tokenization shows that different parts of the same sentence can have different tokenization patterns. Some words are represented by a single token, while other words may be split into multiple tokens.

The `Ġ` marker appearing before several English tokens represents a tokenizer-specific whitespace boundary.

The displayed Arabic token strings use a byte-level representation in the notebook output, so the decoded text and token IDs provide a clearer representation for analysis.

### Finding

Mixed-language text is tokenized according to the tokenizer vocabulary rather than natural word boundaries.

The same tokenizer can process Arabic and English within a single sequence, but tokenization efficiency can vary across languages and individual words.

The number of words is therefore not sufficient to estimate the number of tokens processed by the model. Token count is the relevant measurement for understanding LLM context usage.

## 1.13 Code Tokenization

In [14]:
# Tokenize a code snippet

code = """def add_numbers(a, b):
    return a + b
"""

code_tokens = tokenizer.tokenize(code)
code_ids = tokenizer(code)["input_ids"]

print("Code:")
print(code)

print("Tokens:", code_tokens)
print("Token IDs:", code_ids)

print("Character count:", len(code))
print("Word count:", len(code.split()))
print("Token count:", len(code_ids))

Code:
def add_numbers(a, b):
    return a + b

Tokens: ['def', 'Ġadd', '_numbers', '(a', ',', 'Ġb', '):', 'Ċ', 'ĠĠĠ', 'Ġreturn', 'Ġa', 'Ġ+', 'Ġb', 'Ċ']
Token IDs: [727, 884, 31871, 2784, 11, 292, 1590, 198, 262, 460, 264, 478, 292, 198]
Character count: 40
Word count: 7
Token count: 14


### Result Interpretation

The code snippet contains 40 characters, 7 whitespace-separated words, and 14 tokens.

The tokenizer does not treat programming code as a sequence of complete words. It also represents identifiers, punctuation, operators, whitespace, indentation, and line breaks as part of the token sequence.

For example, the identifier `add_numbers` is split into multiple tokens, while punctuation such as `,` can appear as an individual token.

The `Ċ` marker represents a line break, while the `Ġ` marker indicates a whitespace-related boundary in the tokenizer representation.

The token IDs provide the numerical representation corresponding to these tokenizer outputs.

### Finding

Code tokenization follows the tokenizer vocabulary and segmentation rules rather than programming-language word boundaries.

A short code snippet can therefore produce significantly more tokens than its whitespace-separated word count suggests.

This demonstrates that token count is a more relevant measurement than word count when estimating how much code or text an LLM processes.

## 1.14 Numbers & Punctuation

In [15]:
# Tokenize numbers and punctuation

text = "The price is 12,500.75 USD! Is it 100% correct?"

tokens = tokenizer.tokenize(text)
input_ids = tokenizer(text)["input_ids"]

print("Text:")
print(text)

print("\nTokens:")
print(tokens)

print("\nToken IDs:")
print(input_ids)

print("\nCharacter count:", len(text))
print("Word count:", len(text.split()))
print("Token count:", len(input_ids))

Text:
The price is 12,500.75 USD! Is it 100% correct?

Tokens:
['The', 'Ġprice', 'Ġis', 'Ġ', '1', '2', ',', '5', '0', '0', '.', '7', '5', 'ĠUSD', '!', 'ĠIs', 'Ġit', 'Ġ', '1', '0', '0', '%', 'Ġcorrect', '?']

Token IDs:
[760, 3240, 369, 220, 16, 17, 11, 20, 15, 15, 13, 22, 20, 18954, 0, 2091, 424, 220, 16, 15, 15, 4, 4252, 30]

Character count: 47
Word count: 9
Token count: 24


### Result Interpretation

The input contains 47 characters, 9 whitespace-separated words, and 24 tokens.

The number `12,500.75` is not represented as a single token. It is segmented into multiple tokens containing individual digits and punctuation.

Similarly, `100%` is segmented into separate digit tokens followed by a percentage symbol.

In contrast, several ordinary words are represented by single tokens. The tokenizer therefore applies segmentation rules based on its vocabulary rather than human-defined word or number boundaries.

The `Ġ` marker represents a tokenizer-specific whitespace boundary, while punctuation such as commas, periods, exclamation marks, question marks, and percentage symbols can appear as separate tokens.

### Finding

Numbers and punctuation can produce substantially different tokenization patterns from ordinary words.

A numerical value that appears as one logical unit to a human can require multiple tokens for the tokenizer.

Therefore, word count and visible text structure are not reliable measures of the number of tokens processed by an LLM.

## 1.15 Long Text Tokenization

In [16]:
# Tokenize a longer technical text

long_text = """
Large language models process text through a numerical representation pipeline.
The tokenizer converts the input string into a sequence of tokens, and each token
is mapped to an integer ID from the model vocabulary. These IDs are later used
to retrieve embedding vectors that become inputs to the transformer.

For a local AI assistant, tokenization is especially important because the model
does not receive raw characters directly. Conversation history, user prompts,
system instructions, retrieved documents, and generated content all consume
tokens. As the amount of context increases, the number of tokens processed by
the model also increases.

Understanding token counts is therefore essential for context management,
memory usage, latency analysis, and later inference optimization.
"""

tokens = tokenizer.tokenize(long_text)
input_ids = tokenizer(long_text)["input_ids"]

print("Character count:", len(long_text))
print("Word count:", len(long_text.split()))
print("Token count:", len(input_ids))

print("\nFirst 50 tokens:")
print(tokens[:50])

print("\nFirst 50 token IDs:")
print(input_ids[:50])

Character count: 792
Word count: 114
Token count: 144

First 50 tokens:
['Ċ', 'Large', 'Ġlanguage', 'Ġmodels', 'Ġprocess', 'Ġtext', 'Ġthrough', 'Ġa', 'Ġnumerical', 'Ġrepresentation', 'Ġpipeline', '.', 'Ċ', 'The', 'Ġtokenizer', 'Ġconverts', 'Ġthe', 'Ġinput', 'Ġstring', 'Ġinto', 'Ġa', 'Ġsequence', 'Ġof', 'Ġtokens', ',', 'Ġand', 'Ġeach', 'Ġtoken', 'Ċ', 'is', 'Ġmapped', 'Ġto', 'Ġan', 'Ġinteger', 'ĠID', 'Ġfrom', 'Ġthe', 'Ġmodel', 'Ġvocabulary', '.', 'ĠThese', 'ĠIDs', 'Ġare', 'Ġlater', 'Ġused', 'Ċ', 'to', 'Ġretrieve', 'Ġembedding', 'Ġvectors']

First 50 token IDs:
[198, 33119, 3992, 3983, 1817, 1414, 1472, 264, 33625, 12669, 14855, 13, 198, 760, 44424, 31637, 279, 1879, 886, 1083, 264, 8240, 314, 10885, 11, 321, 1754, 3817, 198, 284, 23098, 310, 449, 7308, 2937, 494, 279, 1558, 33761, 13, 4081, 27437, 513, 2843, 1429, 198, 951, 16672, 37794, 22166]


### Result Interpretation

The long technical text contains 792 characters, 114 whitespace-separated words, and 144 tokens.

The tokenizer therefore produces more tokens than the whitespace-separated word count. For this specific text, the average is approximately 1.26 tokens per word.

The first token is a newline-related token because the input string begins with a line break. This demonstrates that whitespace and line boundaries can contribute to the token sequence.

The first 50 tokens also show that most common English words are represented efficiently, while punctuation and newline characters are represented as part of the tokenizer output.

The observed token count represents the amount of input that would enter the model after tokenization.

### Finding

Longer text increases the number of tokens processed by the model.

Token count is therefore a more meaningful measurement for LLM context usage than character count or word count.

This experiment reinforces the relationship between text length, token count, and the computational resources required to process model input.

## 1.16 Special Tokens

In [17]:
# Inspect special tokens configured for the tokenizer

print("Special tokens map:")
print(tokenizer.special_tokens_map)

print("\nSpecial tokens:")
print(tokenizer.all_special_tokens)

print("\nSpecial token IDs:")
print(tokenizer.all_special_ids)

Special tokens map:
{'eos_token': '<|im_end|>', 'pad_token': '<|endoftext|>', 'audio_bos_token': '<|audio_start|>', 'audio_eos_token': '<|audio_end|>', 'audio_token': '<|audio_pad|>', 'image_token': '<|image_pad|>', 'video_token': '<|video_pad|>', 'vision_bos_token': '<|vision_start|>', 'vision_eos_token': '<|vision_end|>'}

Special tokens:
['<|im_end|>', '<|endoftext|>', '<|audio_start|>', '<|audio_end|>', '<|audio_pad|>', '<|image_pad|>', '<|video_pad|>', '<|vision_start|>', '<|vision_end|>', '<|im_start|>', '<|object_ref_start|>', '<|object_ref_end|>', '<|box_start|>', '<|box_end|>', '<|quad_start|>', '<|quad_end|>', '<|vision_pad|>']

Special token IDs:
[248046, 248044, 248070, 248071, 248076, 248056, 248057, 248053, 248054, 248045, 248047, 248048, 248049, 248050, 248051, 248052, 248055]


### Result Interpretation

The tokenizer configuration contains multiple special tokens with dedicated roles and integer token IDs.

The output identifies an end-of-sequence token, a padding token, and additional tokens associated with audio, image, video, and vision-related processing.

The `special_tokens_map` shows the special tokens assigned to specific tokenizer roles, while `all_special_tokens` contains the complete set of special tokens recognized by the tokenizer.

Each special token has a corresponding integer ID in `all_special_ids`.

These tokens are part of the tokenizer configuration and are distinct from ordinary text tokens.

### Finding

Special tokens provide structured control markers for the model input pipeline.

They have dedicated token IDs and can represent sequence boundaries, padding, or modality-related markers.

The presence of modality-related special tokens in the tokenizer configuration does not by itself demonstrate that multimodal inference has been performed.

## 1.17 Observations

In [18]:
# Summarize the observations from the tokenization experiments

observations = {
    "basic_english_tokens": 5,
    "basic_english_characters": 24,
    "basic_english_words": 4,
    "arabic_tokens": 10,
    "arabic_characters": 30,
    "arabic_words": 5,
    "mixed_language_tokens": 7,
    "mixed_language_characters": 39,
    "mixed_language_words": 5,
    "code_tokens": 14,
    "code_characters": 40,
    "code_words": 7,
    "numbers_punctuation_tokens": 24,
    "numbers_punctuation_characters": 47,
    "numbers_punctuation_words": 9,
    "long_text_tokens": 144,
    "long_text_characters": 792,
    "long_text_words": 114,
}

print("Tokenization Observations")
print("=" * 30)

print(f"Basic English: {observations['basic_english_tokens']} tokens")
print(f"Arabic: {observations['arabic_tokens']} tokens")
print(f"Mixed language: {observations['mixed_language_tokens']} tokens")
print(f"Code: {observations['code_tokens']} tokens")
print(f"Numbers and punctuation: {observations['numbers_punctuation_tokens']} tokens")
print(f"Long text: {observations['long_text_tokens']} tokens")

print("\nKey relationships:")
print("Character count != word count != token count")

print("\nAdditional observations:")
print("- Tokens do not always correspond to complete words.")
print("- Arabic words can be split into multiple tokens.")
print("- Code contains identifiers, punctuation, whitespace, and line breaks.")
print("- Numbers can be split into multiple tokens.")
print("- Longer text produces more tokens.")
print("- Special tokens have dedicated token IDs.")

Tokenization Observations
Basic English: 5 tokens
Arabic: 10 tokens
Mixed language: 7 tokens
Code: 14 tokens
Numbers and punctuation: 24 tokens
Long text: 144 tokens

Key relationships:
Character count != word count != token count

Additional observations:
- Tokens do not always correspond to complete words.
- Arabic words can be split into multiple tokens.
- Code contains identifiers, punctuation, whitespace, and line breaks.
- Numbers can be split into multiple tokens.
- Longer text produces more tokens.
- Special tokens have dedicated token IDs.


### Result Interpretation

The experiments show that tokenization does not follow human-defined units such as words or characters. The tokenizer represents English text, Arabic text, mixed-language text, source code, numbers, punctuation, whitespace, and line breaks according to the token vocabulary and segmentation rules.

The observed token counts were:

- Basic English: 5 tokens
- Arabic: 10 tokens
- Mixed language: 7 tokens
- Code: 14 tokens
- Numbers and punctuation: 24 tokens
- Long text: 144 tokens

The results confirm that character count, word count, and token count are different measurements. A single word may be represented by multiple tokens, while punctuation, whitespace, or parts of numbers may also occupy individual tokens.

The experiments also show that longer input produces more tokens, and special tokens are represented by dedicated token IDs.

### Finding

Token count is the fundamental unit for understanding how text is represented by an LLM. Word count and character count cannot reliably predict the exact number of tokens.

The experiments demonstrate that tokenization behavior depends on the input type and its textual structure. This makes token-level analysis important for context-window management, prompt construction, memory usage, and later inference analysis.

## 1.18 Final Summary

In [19]:
print("Day 1 - Tokenization Final Summary")
print("=" * 40)

summary = [
    "Raw text is converted into tokens by the tokenizer.",
    "Tokens are mapped to integer token IDs.",
    "A token does not necessarily represent a complete word.",
    "Character count, word count, and token count are different measurements.",
    "Arabic and mixed-language text can have different tokenization behavior.",
    "Source code includes identifiers, punctuation, whitespace, and line breaks.",
    "Numbers can be split into multiple tokens.",
    "Longer input generally produces more tokens.",
    "Special tokens have dedicated token IDs.",
]

for index, item in enumerate(summary, start=1):
    print(f"{index}. {item}")

print("\nCore pipeline:")
print("Raw Text -> Tokenizer -> Tokens -> Token IDs")

Day 1 - Tokenization Final Summary
1. Raw text is converted into tokens by the tokenizer.
2. Tokens are mapped to integer token IDs.
3. A token does not necessarily represent a complete word.
4. Character count, word count, and token count are different measurements.
5. Arabic and mixed-language text can have different tokenization behavior.
6. Source code includes identifiers, punctuation, whitespace, and line breaks.
7. Numbers can be split into multiple tokens.
8. Longer input generally produces more tokens.
9. Special tokens have dedicated token IDs.

Core pipeline:
Raw Text -> Tokenizer -> Tokens -> Token IDs


### Result Interpretation

The tokenization experiments established the first stages of the LLM text-processing pipeline. Raw text is converted into tokens, and each token is represented by an integer token ID from the tokenizer vocabulary.

The experiments demonstrated that token boundaries depend on the input structure. English text, Arabic text, mixed-language text, source code, numbers, punctuation, whitespace, and line breaks can produce different tokenization patterns.

The results also confirmed that character count, word count, and token count are different measurements. Longer inputs generally produce more tokens, while special tokens are represented using dedicated token IDs.

### Finding

Tokenization converts human-readable text into a discrete numerical representation that can be processed by the model.

The core pipeline established in this section is:

Raw Text -> Tokenizer -> Tokens -> Token IDs

This token-level representation is the foundation for the next stage of the pipeline: converting token IDs into embedding vectors.

## 1.19 Embedding Lookup

In [20]:
import torch

# A small illustrative vocabulary
vocab_size = 10
embedding_dim = 4

# Create a small embedding matrix
embedding_matrix = torch.arange(
    vocab_size * embedding_dim,
    dtype=torch.float32
).reshape(vocab_size, embedding_dim)

# Example token IDs
token_ids = torch.tensor([2, 5, 7])

# Lookup embeddings using token IDs
embeddings = embedding_matrix[token_ids]

print("Embedding Matrix:")
print(embedding_matrix)

print("\nToken IDs:")
print(token_ids)

print("\nSelected Embeddings:")
print(embeddings)

print("\nEmbedding Matrix Shape:", embedding_matrix.shape)
print("Selected Embeddings Shape:", embeddings.shape)

Embedding Matrix:
tensor([[ 0.,  1.,  2.,  3.],
        [ 4.,  5.,  6.,  7.],
        [ 8.,  9., 10., 11.],
        [12., 13., 14., 15.],
        [16., 17., 18., 19.],
        [20., 21., 22., 23.],
        [24., 25., 26., 27.],
        [28., 29., 30., 31.],
        [32., 33., 34., 35.],
        [36., 37., 38., 39.]])

Token IDs:
tensor([2, 5, 7])

Selected Embeddings:
tensor([[ 8.,  9., 10., 11.],
        [20., 21., 22., 23.],
        [28., 29., 30., 31.]])

Embedding Matrix Shape: torch.Size([10, 4])
Selected Embeddings Shape: torch.Size([3, 4])


### Result Interpretation

The experiment demonstrates the basic embedding lookup mechanism.

The embedding matrix has shape `[10, 4]`, meaning it contains 10 vocabulary entries and each entry is represented by a 4-dimensional vector.

The selected token IDs were `[2, 5, 7]`. Each token ID was used as an index into the embedding matrix:

- Token ID 2 selected row 2: `[8, 9, 10, 11]`
- Token ID 5 selected row 5: `[20, 21, 22, 23]`
- Token ID 7 selected row 7: `[28, 29, 30, 31]`

The resulting tensor has shape `[3, 4]`, because three token IDs were selected and each token is represented by a four-dimensional vector.

The numerical values are illustrative. The important mechanism is the lookup relationship between token IDs and rows of the embedding matrix.

### Finding

A token ID is an index, not an embedding representation.

The embedding process can be summarized as:

Token ID -> Embedding Matrix Row -> Embedding Vector

For a vocabulary size `V` and embedding dimension `D`, the embedding matrix is conceptually represented as `[V, D]`.

## 1.20 Real Token IDs for Embedding Lookup

In [21]:
text = "I love machine learning."

# Get real token IDs from the Qwen tokenizer
input_ids = tokenizer(text)["input_ids"]

# Convert token IDs to a PyTorch tensor
token_ids = torch.tensor(input_ids, dtype=torch.long)

print("Text:")
print(text)

print("\nTokens:")
print(tokenizer.tokenize(text))

print("\nReal Token IDs:")
print(token_ids)

print("\nNumber of Token IDs:", token_ids.numel())
print("Token ID Tensor Shape:", token_ids.shape)

print("\nTokenizer Vocabulary Size:", tokenizer.vocab_size)

Text:
I love machine learning.

Tokens:
['I', 'Ġlove', 'Ġmachine', 'Ġlearning', '.']

Real Token IDs:
tensor([  40, 2854, 5484, 6618,   13])

Number of Token IDs: 5
Token ID Tensor Shape: torch.Size([5])

Tokenizer Vocabulary Size: 248044


### Result Interpretation

The experiment uses the actual Qwen tokenizer to convert the text `"I love machine learning."` into five token IDs:

`[40, 2854, 5484, 6618, 13]`

The token IDs are represented as a PyTorch tensor with shape `[5]`, corresponding to the five tokens produced by the tokenizer.

The tokenizer reports a vocabulary size of `248044`. Each token ID identifies an entry in this vocabulary.

The important distinction is that these integer values are token identifiers, not embedding vectors. In the model pipeline, token IDs are used as indices to retrieve learned embedding vectors from the model's embedding matrix.

### Finding

Real token IDs form the numerical bridge between tokenization and embeddings.

The observed pipeline is:

Raw Text -> Tokens -> Token IDs -> Embedding Lookup

The token IDs identify vocabulary entries, while the embedding lookup converts those IDs into learned numerical vectors inside the pretrained model.

## 1.21 Embedding Dimensions

In [22]:
# Model-side embedding configuration
embedding_dim = 5120
vocab_size = tokenizer.vocab_size

# Use the token IDs from the previous experiment
sequence_length = token_ids.shape[0]

print("Model:", "Qwen3.8-27B")

print("\nVocabulary Size:", vocab_size)
print("Embedding Dimension:", embedding_dim)
print("Sequence Length:", sequence_length)

print("\nConceptual Embedding Matrix Shape:")
print(f"[{vocab_size}, {embedding_dim}]")

print("\nConceptual Input Embedding Shape:")
print(f"[{sequence_length}, {embedding_dim}]")

Model: Qwen3.8-27B

Vocabulary Size: 248044
Embedding Dimension: 5120
Sequence Length: 5

Conceptual Embedding Matrix Shape:
[248044, 5120]

Conceptual Input Embedding Shape:
[5, 5120]


### Result Interpretation

The experiment connects the tokenizer output to the conceptual dimensions of the model's embedding space.

For the Qwen3.8-27B configuration used in this project, the observed values are:

- Vocabulary size: `248044`
- Embedding dimension: `5120`
- Sequence length: `5`

The conceptual embedding matrix therefore has the shape `[248044, 5120]`.

The input text contains five token IDs. After embedding lookup, each token would be represented by a vector with 5120 dimensions. Therefore, the conceptual embedding output for this sequence has shape `[5, 5120]`.

The embedding matrix and output tensor are conceptual in this experiment because the full model weights were not loaded into memory.

### Finding

The embedding stage transforms a sequence of token IDs with shape `[sequence_length]` into a sequence of embedding vectors with shape `[sequence_length, embedding_dimension]`.

For the current example:

Token IDs `[5]` -> Embedding Vectors `[5, 5120]`

This establishes the transition from discrete token identifiers to continuous numerical representations that can be processed by the Transformer.

## 1.22 Embedding Table Size

In [23]:
vocab_size = tokenizer.vocab_size
embedding_dim = 5120

# Number of embedding parameters
embedding_parameters = vocab_size * embedding_dim

# Memory estimates
fp32_bytes = embedding_parameters * 4
fp16_bytes = embedding_parameters * 2

fp32_gb = fp32_bytes / (1024 ** 3)
fp16_gb = fp16_bytes / (1024 ** 3)

print("Embedding Table Analysis")
print("=" * 30)

print("Vocabulary Size:", vocab_size)
print("Embedding Dimension:", embedding_dim)

print("\nEmbedding Parameters:")
print(embedding_parameters)

print("\nEstimated Memory:")
print(f"FP32: {fp32_gb:.2f} GB")
print(f"FP16: {fp16_gb:.2f} GB")

print("\nConceptual Shape:")
print(f"[{vocab_size}, {embedding_dim}]")

Embedding Table Analysis
Vocabulary Size: 248044
Embedding Dimension: 5120

Embedding Parameters:
1269985280

Estimated Memory:
FP32: 4.73 GB
FP16: 2.37 GB

Conceptual Shape:
[248044, 5120]


### Result Interpretation

The embedding table is conceptually represented by the shape `[V, D]`, where `V` is the vocabulary size and `D` is the embedding dimension.

For the current model configuration:

- Vocabulary size: `248044`
- Embedding dimension: `5120`
- Embedding parameters: `1269985280`

The theoretical memory requirements are approximately:

- FP32: `4.73 GB`
- FP16: `2.37 GB`

These values represent the embedding table only. They do not represent the total memory required by the complete Qwen3.8-27B model.

### Finding

The embedding table can contain a very large number of parameters when the vocabulary and embedding dimension are large.

The conceptual relationship is:

Embedding Table Shape = `[Vocabulary Size, Embedding Dimension]`

Token IDs act as indices into this table, and the selected rows provide the embedding vectors that become inputs to the Transformer.

## 1.23 Token IDs → Transformer Input Shape

In [24]:
text = "I love machine learning."

# Get real token IDs from the Qwen tokenizer
input_ids = tokenizer(text)["input_ids"]

# Convert to a PyTorch tensor
token_ids = torch.tensor(input_ids, dtype=torch.long)

# Add a batch dimension
input_ids_batched = token_ids.unsqueeze(0)

# Model configuration used for the conceptual embedding shape
embedding_dim = 5120

print("Text:")
print(text)

print("\nToken IDs:")
print(token_ids)

print("\nToken ID Shape:")
print(token_ids.shape)

print("\nBatched Token IDs:")
print(input_ids_batched)

print("\nBatched Token ID Shape:")
print(input_ids_batched.shape)

print("\nConceptual Transformer Input Shape:")
print(f"[{input_ids_batched.shape[0]}, {input_ids_batched.shape[1]}, {embedding_dim}]")

print("\nDimensions:")
print("Batch Size:", input_ids_batched.shape[0])
print("Sequence Length:", input_ids_batched.shape[1])
print("Embedding Dimension:", embedding_dim)

Text:
I love machine learning.

Token IDs:
tensor([  40, 2854, 5484, 6618,   13])

Token ID Shape:
torch.Size([5])

Batched Token IDs:
tensor([[  40, 2854, 5484, 6618,   13]])

Batched Token ID Shape:
torch.Size([1, 5])

Conceptual Transformer Input Shape:
[1, 5, 5120]

Dimensions:
Batch Size: 1
Sequence Length: 5
Embedding Dimension: 5120


### Result Interpretation

The experiment adds a batch dimension to the real token IDs produced by the Qwen tokenizer.

The original token ID tensor has shape `[5]`, representing five tokens.

After adding a batch dimension, the shape becomes `[1, 5]`:

- `1` represents the batch size.
- `5` represents the sequence length.

Using the model's configured embedding dimension of `5120`, the conceptual embedding output for this input would have shape `[1, 5, 5120]`.

This represents one sequence containing five tokens, where each token is represented by a vector with 5120 dimensions.

The `[1, 5, 5120]` shape is conceptual in this experiment because the actual Qwen3.8-27B embedding weights were not loaded.

### Finding

The token ID sequence is transformed conceptually from:

`[sequence_length]`

to a batched Transformer representation:

`[batch_size, sequence_length, embedding_dimension]`

For the current example:

`[5] -> [1, 5] -> [1, 5, 5120]`

This establishes the tensor shape needed to understand how token embeddings enter the Transformer pipeline.

## 1.24 Why Position Matters

The Transformer needs to know not only which tokens are present, but also where each token appears in the sequence.

For example:

```text
"dog bites man"
"man bites dog"

In [25]:
text_a = "dog bites man"
text_b = "man bites dog"

tokens_a = tokenizer.tokenize(text_a)
tokens_b = tokenizer.tokenize(text_b)

ids_a = tokenizer(text_a)["input_ids"]
ids_b = tokenizer(text_b)["input_ids"]

print("Sequence A:")
print(text_a)
print("Tokens:", tokens_a)
print("Token IDs:", ids_a)

print("\nSequence B:")
print(text_b)
print("Tokens:", tokens_b)
print("Token IDs:", ids_b)

print("\nSame Token Count:", len(ids_a) == len(ids_b))
print("Same Token ID Set:", set(ids_a) == set(ids_b))
print("Same Token ID Sequence:", ids_a == ids_b)

Sequence A:
dog bites man
Tokens: ['dog', 'Ġbites', 'Ġman']
Token IDs: [17907, 46721, 855]

Sequence B:
man bites dog
Tokens: ['man', 'Ġbites', 'Ġdog']
Token IDs: [1461, 46721, 5388]

Same Token Count: True
Same Token ID Set: False
Same Token ID Sequence: False


### Result Interpretation

The experiment compares two sequences containing the same three words but in different orders:

- Sequence A: `dog bites man`
- Sequence B: `man bites dog`

Both sequences contain three tokens, so the token counts are equal.

However, the token IDs are different:

- Sequence A: `[17907, 46721, 855]`
- Sequence B: `[1461, 46721, 5388]`

The middle token, `bites`, remains the same, while the first and third tokens change because their positions contain different words.

The results confirm that:

- `Same Token Count: True`
- `Same Token ID Set: False`
- `Same Token ID Sequence: False`

The sequence order therefore changes the token ID sequence.

### Finding

Token identity alone is not sufficient to represent the structure of a sequence.

The position of each token matters because changing the order of tokens changes the input sequence:

`dog -> bites -> man`

is different from:

`man -> bites -> dog`

This establishes the importance of position information before the sequence is processed by the Transformer.

## 1.25 Decoder Only Transformer Flow

In [26]:
import torch

text = "I love machine learning."

# Real token IDs from the Qwen tokenizer
input_ids = tokenizer(text, return_tensors="pt")["input_ids"]

# Model configuration
embedding_dim = 5120

# Conceptual embedding representation
token_embeddings = torch.zeros(
    input_ids.shape[0],
    input_ids.shape[1],
    embedding_dim,
    dtype=torch.float32
)

# Conceptual Transformer output
hidden_states = torch.zeros_like(token_embeddings)

# Print the pipeline
print("Text:")
print(text)

print("\nInput IDs Shape:")
print(input_ids.shape)

print("\nToken Embeddings Shape:")
print(token_embeddings.shape)

print("\nHidden States Shape:")
print(hidden_states.shape)

print("\nDecoder-Only Transformer Flow:")
print("Token IDs")
print("   ↓")
print("Token Embeddings + Position Information")
print("   ↓")
print("Transformer Blocks")
print("   ↓")
print("Final Hidden States")
print("   ↓")
print("Language Modeling Head")
print("   ↓")
print("Logits")

Text:
I love machine learning.

Input IDs Shape:
torch.Size([1, 5])

Token Embeddings Shape:
torch.Size([1, 5, 5120])

Hidden States Shape:
torch.Size([1, 5, 5120])

Decoder-Only Transformer Flow:
Token IDs
   ↓
Token Embeddings + Position Information
   ↓
Transformer Blocks
   ↓
Final Hidden States
   ↓
Language Modeling Head
   ↓
Logits


### Result Interpretation

The experiment connects the token ID representation to the high-level decoder-only Transformer pipeline.

The input text is converted into token IDs with shape `[1, 5]`:

- `1` represents the batch size.
- `5` represents the sequence length.

The conceptual token embedding representation has shape `[1, 5, 5120]`:

- `1` represents the batch size.
- `5` represents the sequence length.
- `5120` represents the embedding dimension.

The conceptual hidden-state representation also has shape `[1, 5, 5120]`.

The high-level decoder-only Transformer flow is:

Token IDs
-> Token Embeddings + Position Information
-> Transformer Blocks
-> Final Hidden States
-> Language Modeling Head
-> Logits

The embedding and hidden-state tensors in this experiment are conceptual representations rather than outputs from the loaded Qwen3.8-27B model weights.

### Finding

The experiment establishes the main data flow from discrete token IDs to continuous representations processed by the Transformer.

The key tensor transition is:

`[batch_size, sequence_length]`

to:

`[batch_size, sequence_length, embedding_dimension]`

The Transformer then processes these representations through its blocks before producing final hidden states that are passed to the language modeling head.